src/wast-parser.cc

```diff
@@ -1732,50 +1732,139 @@ Result WastParser::ParsePlainInstr(std::unique_ptr<Expr>* out_expr) {
+  Result WastParser::ParseSimdV128Const(Const* const_, TokenType token_type) {
+  WABT_TRACE(ParseSimdV128Const);
 
-        // Expected one 0xXXXXXXXX number
-        if (!PeekMatch(TokenType::Nat))
-          return ErrorExpected({"an Nat literal"}, "123");
+  switch (token_type) {
+    case TokenType::Nat:
+    case TokenType::Int:
+    case TokenType::Float:
+    case TokenType::ValueType: {
+      Error(
+        const_->loc,
+        "Unexpected type at start of simd constant. "
+        "Expected one of: i8x16, i16x8, i32x4, i64x2, f32x4, f64x2."
+      );
+      return Result::Error;
+    }
+    default: break;
+  }
+
+  auto text = Consume().text();
+  // Look up the number of lanes and whether the v128 is an integer or
+  // floating-point vector:
+  uint8_t no_lanes = 0;
+  bool integer = true;
+  if (text == "i8x16") {
+    no_lanes = 16;
+  } else if (text == "i16x8") {
+    no_lanes = 8;
+  } else if (text == "i32x4") {
+    no_lanes = 4;
+  } else if (text == "i64x2") {
+    no_lanes = 2;
+  } else if (text == "f32x4") {
+    no_lanes = 4;
+    integer = false;
+  } else if (text == "f64x2") {
+    no_lanes = 2;
+    integer = false;
+  } else {
+    Error(
+      const_->loc,
+      "Unexpected type at start of simd constant. "
+      "Expected one of: i8x16, i16x8, i32x4, i64x2, f32x4, f64x2."
+    );
+    return Result::Error;
+  }
+  uint8_t lane_size = sizeof(v128) / no_lanes;
 
-        Literal literal = Consume().literal();
+  // The bytes of the v128 are written here first:
+  std::array<char, 16> v128_bytes{};
+  const_->loc = GetLocation();
 
-        string_view sv = literal.text;
-        const char* s = sv.begin();
-        const char* end = sv.end();
-        Result result;
+  for (int i = 0; i < no_lanes; ++i) {
+    Location loc = GetLocation();
 
-        result = ParseInt32(s, end, &(const_->v128_bits.v[i]),
-                            ParseIntType::SignedAndUnsigned);
+    // Check that the lane literal type matches the element type of the v128:
+    if (integer) {
+      if (!PeekMatch(TokenType::Nat)) {
+        return ErrorExpected({"a Nat literal"}, "123");
+      }
+    } else {
+      if (!PeekMatch(TokenType::Float)) {
+        return ErrorExpected({"a Float literal"}, "42.0");
+      }
+    }
 
-        if (Failed(result)) {
-          Error(loc, "invalid literal \"%s\"", literal.text.c_str());
-          return Result::Error;
+    Literal literal = Consume().literal();
+
+    string_view sv = literal.text;
+    const char* s = sv.begin();
+    const char* end = sv.end();
+
+    // Pointer to the lane in the v128 bytes:
+    char* lane_ptr = &v128_bytes[lane_size * i];
+    Result result;
+
+    // For each type, parse the next literal, bound check it, and write it to
+    // the array of bytes:
+    if (integer) {
+      switch(no_lanes) {
+        case 16: {
+          uint8_t value[4] = {0, 0, 0, 0};
+          result = ParseInt32(s, end, Bitcast<uint32_t*>(&value), ParseIntType::SignedAndUnsigned);
```

Comment:
I think it might be easier to parse this as a `uint32_t`, then convert to `int32_t` and see if it is in the range `[-128,127]`.

Reply:
Yeah, this sounds good to me :D

Reply:
So I think I've fixed this. Two changes were needed: 

* parsing these as both Int and Nat (since Nat does not support negative numbers)
* checking the range [signed::MIN, unsigned::MAX], otherwise 255 fails. 

I've added tests for the corner cases, except for i64x2 where the test suite does not appear to support the largest 64-bit values yet.